File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.py
class Solution:
    def totalCost(self, cost: List[int], n: int, memo: dict[int,int]) -> int:
        if (n < 0):
            return 0
        if (n == 0 or n == 1):
            return cost[n]

        if n not in memo:
            memo[n] = cost[n] + min(self.totalCost(cost, n-1, memo), self.totalCost(cost, n-2, memo))
        return memo[n]


    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n = len(cost)
        memo = {}
        return min(self.totalCost(cost, n-1, memo), self.totalCost(cost, n-2, memo))

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */

var totalCost = function(cost, n, memo){
    if (n in memo) return memo[n];
    if (n<0) return 0;
    if (n === 0 || n === 1){
        return cost[n];
    }

    memo[n] = cost[n] + Math.min(totalCost(cost, n-1, memo), totalCost(cost, n-2, memo));
    return memo[n];
}
var minCostClimbingStairs = function(cost) {
    const n = cost.length;
    let memo = {};
    return Math.min(totalCost(cost, n-1, memo), totalCost(cost, n-2, memo));
};